Grid: properly re-render when the `layout` prop changes dynamically.

retrofox opened the pull request:
Related to ARC-1330

## Why?
The Grid component wasn't responding to external layout changes. When the layout prop was updated (e.g., adding/removing tiles or switching to a different layout), the Grid would retain old tile instances instead of updating to reflect the new layout.

This forced users to use workarounds like adding a `key` prop to force re-mounting.

## How?
**Root cause:** The `temporaryLayout` state was initialized with the `layout` prop value but never synchronized when the prop changed externally.

**Solution:** Remove the initial value from `temporaryLayout` state, so it starts as `undefined` and only gets populated during drag/resize operations:

This ensures:
- `temporaryLayout` only exists during active drag/resize operations
- When `layout` prop changes externally, `activeLayout` immediately uses the new value
- No functionality is lost (there's no cancel/revert feature)

## Testing
Added an interactive Storybook story (`InteractiveGrid`) that demonstrates:
- Adding tiles dynamically
- Removing tiles individually
- Switching between different layouts
- Edit mode with drag & drop and resize

Before the fix, these operations wouldn't update the grid. After the fix, all dynamic layout changes work correctly.

## Additional Changes
- Improved the interactive story with edit mode toggle and per-tile removal
- Fixed typo in comment: "updaing" → "updating"

https://github.com/user-attachments/assets/e5ee3644-66a8-48f4-9aec-1c6595714bd1

retrofox committed:
add story demonstrating grid layout change bug
shows tiles not updating when layout prop changes dynamically

retrofox committed:
fix grid not re-rendering when layout prop changes
removes initial value from temporaryLayout state to allow dynamic updates

retrofox committed:
improve interactive grid story to demonstrate dynamic layout features
rename from DynamicLayoutChange to InteractiveGrid, add edit mode toggle and tile removal

retrofox committed:
hide tile remove button when edit mode is disabled

youknowriad commented on packages/grid/src/grid.tsx:
```diff
@@ -17,11 +17,9 @@ export function Grid( {
-	const activeLayout = temporaryLayout || layout;
+	// Temporary layout to avoid updating the layout while dragging
+	const [ temporaryLayout, setTemporaryLayout ] = useState< GridLayoutItem[] | undefined >();
+	const activeLayout = temporaryLayout ?? layout;
```
This make sense, this means that we continue to use the prop value until the user makes a change. Now what this suggests as well is that we should do `setTemporaryLayout( undefined )` whenever we call `onChangeLayout` to avoid issues afterwards.

retrofox replied:
Good call 💯 

Added a useEffect to reset temporaryLayout whenever the layout prop changes externally. This ensures any pending temporary state gets cleared when the parent updates the layout.

https://github.com/Automattic/wp-calypso/commit/e9c99014f5b96eae24f945b9730a43eba94c08d6

youknowriad replied:
I was thinking the opposite calling `setTemporaryLayout( undefined )` at the same time we call `onChangeLayout` instead of when the layout changes externally. 

I don't recall exactly why we use `temporaryLayout` for but if it's for user edits or things like that, I think these would need to have more priority than external changes (when these user changes are not persisted, and they get persisted when you call `onChangeLayout`).

retrofox replied:
It effectively [cleans the temporary layout](https://github.com/Automattic/wp-calypso/blob/7fa3afff80927454d59db894c38317b0f8cbb1ff/packages/grid/src/grid.tsx#L132) when `onChangeLayout` is called (in trunk).

It indeed has higher priority than external changes when they are _internal_, for instance, when dragging or resizing tiles.

Anyway, I've removed code that cleans the temporary layout in the side effect hook. It handled an edge case when the user interacts with the UI while the parent changes the layout, for instance, when the user is dragging or resizing a tile. Although it's possible (very improbable), I think it's the consumer's responsibility to handle it.

retrofox commented:
Good call! Added a `useEffect` to reset `temporaryLayout` whenever the `layout` prop changes externally. This ensures any pending temporary state gets cleared when the parent updates the layout.

https://github.com/Automattic/wp-calypso/commit/e9c99014f5b96eae24f945b9730a43eba94c08d6

retrofox committed:
improve comments explaining temporary layout state management

retrofox committed:
remove side effect to clean tmp layout
It handles an edge case when the user interacts with the UI while the parent changes the layout, for instance, when the user is dragging or resizing a tile. Also, although it's possible (very improbable), it's the consumer's responsibility to handle it.
